Multiple new features added (1.12.2-forge)

Danielxs01 opened the pull request:
### Changelog:

 

1. Added [Packet].sendToPlayer([Player]);  `Clean method for sending packet from server to player`
2. Added [Command].getPlayer(): Optional<Player> `Method allowing access to the command executor on execute(...)`
3. Added [Command].getRequiredPermissionLevel() `Allows to explicitly set the required permission.`
4. Added Command.PermissionLevel `List of available values for [Command].getRequiredPermissionLevel()`

Danielxs01 committed:
Added Packet.sendToPlayer(Player)

Danielxs01 committed:
Changed Permissions and added getPlayer(): Optional<Player>

Danielxs01 committed:
Update README.md

Danielxs01 committed:
Update README.md

Danielxs01 committed:
Update README.md

Danielxs01 committed:
Update AUTHORS

cam72cam commented:
I'll have to take a look at the implementation on the other branches, but I'm leaning toward adding player as a Optional/Nullable argument to execute instead of adding a field to Command.

Danielxs01 commented:
> I'll have to take a look at the implementation on the other branches, but I'm leaning toward adding player as a Optional/Nullable argument to execute instead of adding a field to Command.

The reason why I've done this was backwards compatability, but sure, you could just put it as an argument in execute.
I've done the changes in a way that, if you don't want to use them there is no other work than just upgrading to the new UMC version.

Danielxs01 committed:
Update README.md

cam72cam commented:
I'm OK with making small breaking changes like this, as long as I remember to update the version accordingly.  The next release will be a minor breaking release a 0.X++.0 change due the the rendering anyways.

cam72cam commented:
M.N.O
M: Major breaking change, will need serious refactoring in dependent projects (I avoid if at all possible)
N: Minor breaking change, may need to add an argument or update the usage of a class
O: Backwards compatible changes that either fix a bug or introduce a feature

Danielxs01 commented:
Okay. So execute would be something like:
execute(Optional < Player > player, String[] args){}
Do you want to make that change or should I do that?

cam72cam commented:
`execute(World world, Consumer<PlayerMessage> sender, Optional<Player> player, String[] args);`

The `Consumer<PlayerMessage>` can also be a console or other targets and we should preserve that.

I'd like you to take a stab at it, if you don't mind.

Danielxs01 commented:
> `execute(World world, Consumer<PlayerMessage> sender, Optional<Player> player, String[] args);`
> 
> The `Consumer<PlayerMessage>` can also be a console or other targets and we should preserve that.
> 
> I'd like you to take a stab at it, if you don't mind.

Sure, no problem. 
Yeah, I forgot Consumer<PlayerMessage>..
But do you need the world as an argument? Console always returns the overworld

cam72cam commented:
Hmm, that's a good question.  I think you are right, there are no reasonable scenarios where you could have a command sent with a specific world without a player.  I'm OK with dropping the param.

Danielxs01 commented:
Okay, so its:
execute(Consumer < PlayerMessage > sender,  Optional < Player > player, String[] args);
Sounds good to me 👍

cam72cam commented:
:+1: thanks for working with me on this!

Danielxs01 committed:
Modified execute(...) to execute(Consumer<PlayerMessage> sender,Optional<Player> player,String[] args)
Removed getPlayer(): Optional<Player>
Removed opRequired(): boolean
Modified getRequiredPermissionLevel()

Danielxs01 committed:
Removed opRequired(): boolean
Added getRequiredPermissionLevel(): int
Modified execute() - Now also available for console

Danielxs01 commented:
I'm done with my changes. :)

cam72cam commented on src/main/java/cam72cam/mod/util/ModCoreCommand.java:
```diff
@@ -1,15 +1,19 @@
+import java.util.Optional;
+import java.util.function.Consumer;
+
+import org.codehaus.plexus.util.StringUtils;
```
Would you mind switching this over to apache's string utils?  I try to minimize the number of libraries UMC directly depends upon (even if present in MC)

http://commons.apache.org/proper/commons-lang/apidocs/org/apache/commons/lang3/StringUtils.html#isNumeric-java.lang.CharSequence-

Danielxs01 replied:
Yeah, sure, no problem

Danielxs01 replied:
Kinda went under my radar.
I thought i used the Apache StringUtils already..

Danielxs01 committed:
Changed StringUtils import
Changed from org.codehaus.plexus.util.StringUtils
          to org.apache.commons.lang3.StringUtils

Danielxs01 commented:
Import of StringUtils is fixed now.
